lib/control.ml

```diff
@@ -150,40 +139,90 @@ implicit module Option = struct
     | None -> acc
     | Some x -> f x acc
 
-  let traverse (type a) (type b) {F : Applicative} (f : a -> b F.t)  : a option -> b t F.t = function
+  (* Traversable *)
+  let traverse (type a) (type b) {F : Applicative} (f : a -> b F.t)  : a option -> b option F.t = function
     | None -> F.return None
     | Some x -> F.fmap (fun x -> Some x) (f x)
 end
 
-implicit module List = struct
+implicit module List : sig
+  include Functor with type 'a t = 'a list
+  include Applicative with type 'a t := 'a t
+  include Monad with type 'a t := 'a t
+  include Monad_plus with type 'a t := 'a t
+  include Foldable with type 'a t := 'a t
+  include Traversable with type 'a t := 'a t
+end = struct
   type 'a t = 'a list
 
   (* Functor *)
   let fmap = List.map
 
-  (* Monad *)
+  (* Applicative *)
   let return x = [x]
-  let bind x f =
-    let rec aux acc = function
-      | x :: xs -> aux (x @ acc) xs
-      | [] -> acc in
-    aux [] (List.rev_map f x)
+  let apply fs xs =
+    List.concat (List.map (fun f -> List.map (fun x -> f x) xs) fs)
 
-  (* Applicative *)
-  let pure x = [x]
-  let apply fs xs = bind fs (bind xs)
+  (* Monad *)
+  let bind x f = List.concat (List.map f x)
 
   (* Monad_plus *)
   let mzero = []
   let mplus = (@)
 
   (* Foldable *)
-  let rec fold f t acc = match t with
-    | [] -> acc
-    | x :: xs -> fold f xs (f x acc)
+  let fold f xs a = List.fold_left (fun x y -> f y x) a xs
 
   (* Traversable *)
   let traverse {F : Applicative} f t =
-    let cons x ys = F.apply (F.apply (F.return (fun x xs -> x :: xs)) (f x)) ys in
-    fold cons t (F.return [])
+    let cons x ys = F.apply (F.fmap (fun x xs -> x :: xs) (f x)) ys in
+    List.fold_right cons t (F.return [])
+end
+
+implicit module Function {A : Any} : sig
+  include Functor with type 'b t = A.t_for_any -> 'b
+  include Applicative with type 'b t := 'b t
+  include Monad with type 'b t := 'b t
+end = struct
+  type 'b t = A.t_for_any -> 'b
+
+  (* Functor *)
+  let fmap m f x = m (f x)
+
+  (* Applicative *)
+  let return x _ = x
+  let apply f g x = f x (g x)
+
+  (* Monad *)
+  let bind g f x = f (g x) x
+end
+(** (a -> b) is an instance of Monad b - it behaves like the reader monad *)
+
+implicit module Pair {A : Any} : Functor with type 'b t = A.t_for_any * 'b = struct
+  type 'b t = A.t_for_any * 'b
+
+  let fmap m (a, b) = (a, m b)
+end
+
+type ('a, 'b) const = Const of 'a
```

Comment:
I wonder whether this can be defined simply as
```suggestion
type ('a, 'b) const = 'a
```

If it does need to be a proper (generative) type, it'd be better as a record, so that it's easy to destruct as well as construct:

```ocaml
type ('a, 'b) const = { const: 'a }
```

Reply:
Daniel wants me to merge this so I'm going to make this an issue and deal with it later